core/src/test/scala/org/apache/spark/storage/BlockManagerSuite.scala

```diff
@@ -2266,6 +2270,160 @@ class BlockManagerSuite extends SparkFunSuite with Matchers with PrivateMethodTe
     }
   }
 
+  test("SPARK-41497: getOrElseUpdateRDDBlock do compute based on cache visibility statue") {
+    val store = makeBlockManager(8000, "executor1")
+    val blockId = RDDBlockId(rddId = 1, splitIndex = 1)
+    var computed: Boolean = false
+    val data = Seq(1, 2, 3)
+    val makeIterator = () => {
+      computed = true
+      data.iterator
+    }
+
+    // Cache doesn't exist and is not visible.
+    assert(store.getStatus(blockId).isEmpty && !store.isRDDBlockVisible(blockId))
+    val res1 = store.getOrElseUpdateRDDBlock(
+      1, blockId, StorageLevel.MEMORY_ONLY, classTag[Int], makeIterator)
+    // Put cache successfully and reported block task info.
+    assert(res1.isLeft && computed)
+    verify(master, times(1)).updateRDDBlockTaskInfo(blockId, 1)
+
+    // Cache exists but not visible.
+    computed = false
+    assert(store.getStatus(blockId).nonEmpty && !store.isRDDBlockVisible(blockId))
+    val res2 = store.getOrElseUpdateRDDBlock(
+      1, blockId, StorageLevel.MEMORY_ONLY, classTag[Int], makeIterator)
+    // Load cache successfully and reported block task info.
+    assert(res2.isLeft && computed)
+    verify(master, times(2)).updateRDDBlockTaskInfo(blockId, 1)
```

Comment:
Note:
This specific check is not very accurate ... it is the current code's implementation detail. and not the behavior we want to test for. (I am fine with leaving this check here for now though !)

For example, we can skip reporting to master if the taskId has computed a particular block already (especially if it is in the invisible set). In actual spark usage, this cannot happen - and so this is an artifact of test cases IMO.

A better check would be to test for visibility is false, and computed == true (which we do already) - which is the behavior we want to test for.

Reply:
HI @mridulm actually this is to make sure that task2blockId information will be reported only when the block is invisble as this is part of the work flow design. If this doesn't make sense, I can remove `verify` statement in a new iteration. Thanks.

Reply:
That makes sense, thanks for clarifying !